### DISCUSSION TITLE
Most reliable way to use ui.input

### DISCUSSION THREAD

**Discussion by nigelsim** [zauberzeug/nicegui#2984] [Q&A] [answered]
### Question

I've been using Nice GUI on more and more internal projects lately, but have hit a showstopper relating to the behaviour of `ui.input`, which is reminiscent of https://github.com/zauberzeug/nicegui/issues/287, where the binding is out of date.

My question is, what is the correct way to bind and use data from input elements, where we also want to set default values?

The exact example is this:

```python
connection = {
    "mqtt_host": "rabbitmq",
    "mqtt_port": 1883,
    "mqtt_username": "vhost1:user1",
    "mqtt_password": "password1",
}
ui.input(label="MQTT Host").bind_value(
    connection, "mqtt_host"
)  # dev-diab-api.flyfreely.io
ui.input(label="MQTT Port").bind_value(connection, "mqtt_port")
ui.input(label="MQTT Username").bind_value(
    connection, "mqtt_username"
)
ui.input(label="MQTT Password").bind_value(
    connection, "mqtt_password"
)
ui.button(
    "Connect",
    on_click=lambda: manager.connect(
        connection["mqtt_host"],
        connection["mqtt_port"],
        connection["mqtt_username"],
        connection["mqtt_password"],
    ),
)
```

If I try to change the value of the host from `rabbitmq` to `localhost` and then press `Connect` I'll quite reliably get `localho`. If I then put an extra space at the end and then remove it I'll get the correct string.

I understand the basic two way data sync that is occurring, but regardless, the UI and Python data should be in sync.

**Comment by falkoschindler** [comment]
Hi @nigelsim,

Binding an input field to a dictionary the way you did should work without problems. And I can't reproduce any problem. Even after rapidly typing dozens of characters, the printed value is identical to the one in the input field.

```py
connection = {"mqtt_host": "rabbitmq", "mqtt_port": 1883, "mqtt_username": "vhost1:user1", "mqtt_password": "password1"}
ui.input(label="MQTT Host").bind_value(connection, "mqtt_host")
ui.input(label="MQTT Port").bind_value(connection, "mqtt_port")
ui.input(label="MQTT Username").bind_value(connection, "mqtt_username")
ui.input(label="MQTT Password").bind_value(connection, "mqtt_password")
ui.button("Connect", on_click=lambda: print(connection))
```

Are you using the most recent version 1.4.23? Does it occur with a specific browser? Is there any other code involved, like CPU-intense computations in the background blocking the main thread?

**Reply by nigelsim** [reply]
Hi @falkoschindler, I am on the latest version, using Chromium on Linux.

I've created a minimal project that reproduces it [here](https://github.com/nigelsim/nicegui-issue), and included a video. Note, if I type fast this happens, regardless of if there is a delay at the end of the typing. If I type slowly (1 second between key strokes) it does not happen.

The component that does this is a `@ui.refreshable` so I've included this in the example to be as true to the actual situation, in case that is part of it.

**Reply by falkoschindler** [reply]
Thanks, @nigelsim. Unfortunately, I can't reproduce the problem with this project either.

> I've created a minimal project [...]. The component that does this is a `@ui.refreshable` so I've included this in the example

Well, if the problem occurs without the refreshable (like with [your code snippet](https://github.com/zauberzeug/nicegui/discussions/2984#discussion-6580764) or [mine](https://github.com/zauberzeug/nicegui/discussions/2984#discussioncomment-9263552)), the minimal project doesn't need it.

Can you confirm that the problem occurs with this minimal code?
```py
connection = {'host': ''}
ui.input().bind_value(connection, 'host')
ui.button('Show', on_click=lambda: ui.notify(connection['host']))
```

**Reply by nigelsim** [reply]
It certainly does.

This is a screenshot of the page, and the websocket calls back to the server (which is how I expect the data is kept in sync)
![image](https://github.com/zauberzeug/nicegui/assets/2869/e2a97b53-214a-49b9-be21-461acf171153)

And of the toast
![image](https://github.com/zauberzeug/nicegui/assets/2869/ad1ea2d7-0b20-4366-baba-b08a09f3e867)

**Reply by falkoschindler** [reply]
Very strange... Can anyone else reproduce this problem?

**Reply by frankvp11** [reply]
I can't.

**Reply by nigelsim** [reply]
I can't reproduce it if I use Firefox, and it sends a WS message for every keystroke. 

![image](https://github.com/zauberzeug/nicegui/assets/2869/daaeda94-56e8-45fb-9eaf-f59837cb6102)


But it happens on Chromium 123.0.6312.122

**Comment by nigelsim** [comment, marked as answer]
OK, this was actually the [Web Component Dev Tools](https://chrome.google.com/webstore/detail/gdniinfdlmmmjpnhgnkmfpffipenjljo) Chrome extensions overriding some JS.

I put some logging in the event handling code in `nicegui.js:154`. This received everything correctly, the call `throttle`. ended up in the code for Web Component Dev Tools.

I should have tested in Incognito mode first, but this is the only time I've had an extension do something like this, so I didn't think of it.

Sorry for the run around.

**Reply by nigelsim** [reply]
PS, I've raised an issue on that project.

**Reply by nigelsim** [reply]
Also, I do wonder if it should have been possible for this to occur. Shouldn't the content of nicegui.js be encapsulated somehow so this doesn't occur?

**Reply by falkoschindler** [reply]
@nigelsim Hm, I'm not sure. What do you mean with "encapsulated"? How could we improve nicegui.js? And what exactly is the problem with the current implementation? Thanks!

**Reply by nigelsim** [reply]
`nicegui.js` has all of its functions at the top level, which means they are all attached to `window`. e.g., 

![image](https://github.com/zauberzeug/nicegui/assets/2869/db2c159f-c9d1-4e2b-bb7a-eb30cc16799a)

which means that any other JS import can just overwrite those methods and change the application behaviour, which is exactly what happened to me.

Typically, I've see JS applications just encapsulate their own code inside a self executing block, so that this can't happen.

e.g., 

```javascript
(function() {
const True = true;
const False = false;
const None = undefined;

let app = undefined;
let mounted_app = undefined;

const loaded_libraries = new Set();
const loaded_components = new Set();

function parseElements(raw_elements) {
  return JSON.parse(
    raw_elements
      .replace(/&#36;/g, "$")
      .replace(/&#96;/g, "`")
      .replace(/&gt;/g, ">")
      .replace(/&lt;/g, "<")
      .replace(/&amp;/g, "&")
  );
}

function getElement(id) {
  const _id = id instanceof HTMLElement ? id.id : id;
  return mounted_app.$refs["r" + _id];
}
...
})();
```

Normally the JS bundler like Webpack does this for you, and maybe there is a more modern way to achieve this. 

But having the `window` namespace populated with anything other than config (like you have to do to when setting up embedded widgets like Mixpanel or Hubspot) is unusual.

Of course, it's probably more complicated than just doing this to get the NiceGUI functions off `window` as there are probably a lot of implicit assumptions in the code given that this is how it currently works. Perhaps exposing all of the require data/functions onto `window.NiceGUI` so it is less likely to conflict?

Some links that may or may not be helpful:

https://developer.mozilla.org/en-US/docs/Glossary/IIFE
https://www.codeproject.com/Articles/757949/JavaScript-Encapsulation-using-Anonymous-Functions

**Reply by falkoschindler** [reply]
Thanks, @nigelsim! That makes sense. We'll look into it and try to move (most of) our code out of the global `window` context.

Using bundlers like webpack (or rollup?) is also something we're currently thinking about. On one hand they could improve how NiceGUI serves nicegui.js as well as Vue, Quasar and Tailwind. On the other hand it could lead to a recipe for integrating third-party dependencies into NiceGUI (see PR #2991). But the whole world of NPM and JS bundlers is quite new to me and I'm still trying to get my head around how to use it in the context of NiceGUI.

**Reply by falkoschindler** [reply]
See feature request https://github.com/zauberzeug/nicegui/discussions/3007.
